### PR TITLE
Example/example re assembling fields

### PR DESCRIPTION
**Pull request by luisaFelixSalles**
This new example shows how to do some matrix operations, including basic mathematical operation (power, add and multiply by
a constant, add field containers and invert ) and separating and assembling fields and fields containers

### CODE REVIEW

**Comment by rafacanton** [comment]
@luisaFelixSalles Thanks for the example. I think it would be good to have a follow-up on this one to explain users how to create their own matrix fields "by hand". Dpf Matrix Fields are created "by column" and not "by row", and we are not documenting it anywhere. This also affects the way the generalized_dot_product operator is working.

@PProfizi This caused a bug in the past, so I think it is something we would like to properly document. In addition, it works differently in numpy, so it is not very straigthforward to use without an example

**Comment by PProfizi** [comment]
> @luisaFelixSalles Thanks for the example. I think it would be good to have a follow-up on this one to explain users how to create their own matrix fields "by hand". Dpf Matrix Fields are created "by column" and not "by row", and we are not documenting it anywhere. This also affects the way the generalized_dot_product operator is working.
> 
> @PProfizi This caused a bug in the past, so I think it is something we would like to properly document. In addition, it works differently in numpy, so it is not very straigthforward to use without an example

@luisaFelixSalles I think this remark could go both in a specific tutorial on Field creation, and in the User Guide documentation.

File: examples/01-mathematical-operations/matrix-operations.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
+"""
+.. _ref_matrix-operations:
+
+Matrix Operations
+~~~~~~~~~~~~~~~~~
+
+This example shows how to do some matrix operations, including basic mathematical operation (power, add and multiply by
+a constant, add field containers and invert ) and separating and assembling fields and fields containers
```
Comment (PProfizi) [inline examples/01-mathematical-operations/matrix-operations.py:8]:
```suggestion
This example shows how to do some matrix operations, including basic mathematical operations (power, add and multiply by
a constant, add field containers and invert) and separating and assembling fields and fields containers.
```

File: examples/01-mathematical-operations/matrix-operations.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+"""
+
+###############################################################################
+# Import the ``ansys.dpf.core`` module, included examples file, and the ``DpfPlotter``
+# module.
```
Comment (PProfizi) [inline examples/01-mathematical-operations/matrix-operations.py:14]:
```suggestion
# Import the ``ansys.dpf.core`` module, included examples file, and the ``operators.math``
# module.
```

File: examples/01-mathematical-operations/matrix-operations.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+from ansys.dpf import core as dpf
+from ansys.dpf.core import examples
+import ansys.dpf.core.operators.math as maths
+
+###############################################################################
+# Open an example and print the ``Model`` object. Here a result file from a crankshaft
```
Comment (PProfizi) [inline examples/01-mathematical-operations/matrix-operations.py:20]:
```suggestion
# Load an example and print the ``Model`` object. Here a result file from a crankshaft
```

File: examples/01-mathematical-operations/matrix-operations.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+# Printing the model displays this metadata:
+#
+# - Analysis type
+# - Available results
+# - Size of the mesh
+# - Number of results
+#
+my_model = dpf.Model(examples.find_complex_rst())
+my_mesh = my_model.metadata.meshed_region
+# print(my_model)
```
Comment (PProfizi) [inline examples/01-mathematical-operations/matrix-operations.py:35]: Why comment the print?

File: examples/01-mathematical-operations/matrix-operations.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+###############################################################################
+# Get the stress tensor and define it's scoping. Here, only three nodes will be take into account to facilitate the
+# results visualisation
```
Comment (PProfizi) [inline examples/01-mathematical-operations/matrix-operations.py:38]:
```suggestion
# Get the stress tensor and define its scoping. Only three nodes will be taken into account to facilitate the
# visualization.
```

File: examples/01-mathematical-operations/matrix-operations.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+my_nodes_scoping = dpf.Scoping(ids=[38, 37, 36], location=dpf.locations.elemental)
+my_stress = my_model.results.stress(mesh_scoping=my_nodes_scoping).eval()
+
+# Here we need to average the result from 'elemental_nodal' to an 'elemental' location to
+# facilitate the visualisation of the plot
```
Comment (PProfizi) [inline examples/01-mathematical-operations/matrix-operations.py:43]:
```suggestion
# We need to average the result from 'elemental_nodal' to an 'elemental' location to plot it.
```

File: examples/01-mathematical-operations/matrix-operations.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+my_avg_stress = dpf.operators.averaging.to_elemental_fc(fields_container=my_stress, mesh=my_mesh).eval()
```
Comment (PProfizi) [inline examples/01-mathematical-operations/matrix-operations.py:44]:
```suggestion
my_avg_stress = my_model.results.stress(mesh_scoping=my_nodes_scoping).on_location(dpf.locations.elemental).eval()
```
You can also directly request the result as "elemental" like shown above

File: examples/01-mathematical-operations/matrix-operations.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+# print(my_avg_stress, my_avg_stress[0])
```
Comment (PProfizi) [inline examples/01-mathematical-operations/matrix-operations.py:45]: Why comment the print?

File: examples/01-mathematical-operations/matrix-operations.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+#########################################################
+# Separating tensor by component
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+# If operations need to be done separetly in each tensor component, the
```
Comment (PProfizi) [inline examples/01-mathematical-operations/matrix-operations.py:51]:
```suggestion
# If operations need to be done separately in each tensor component, use
```

File: examples/01-mathematical-operations/matrix-operations.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+# :func:'select_component()<ansys.dpf.core.fields_container.FieldsContainer.select_component>'.
+# Here, the stress tensor has 6 components by elementary data (symmetrical tensor XX,YY,ZZ,XY,YZ,XZ).
```
Comment (PProfizi) [inline examples/01-mathematical-operations/matrix-operations.py:53]:
```suggestion
# Here, the stress tensor has 6 components per elementary data (symmetrical tensor XX,YY,ZZ,XY,YZ,XZ).
```

File: examples/01-mathematical-operations/matrix-operations.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+
+for i in range(0, 6):  # Separating the results in different fields containers for each stress tensor component
+    globals()[f'stress_{i + 1}'] = my_avg_stress.select_component(i)
```
Comment (PProfizi) [inline examples/01-mathematical-operations/matrix-operations.py:56]: We probably lack an operator to do this in one go, which would return a FieldsContainer labeled by 'component'.

File: examples/01-mathematical-operations/matrix-operations.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+
+################################################################################
+# Mathematical operation on each field
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+# Here we will do some basic mathematical operations on
+# Power
```
Comment (PProfizi) [inline examples/01-mathematical-operations/matrix-operations.py:63]: ?

File: examples/01-mathematical-operations/matrix-operations.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+# Compute the power operation for each elementary data
```
Comment (PProfizi) [inline examples/01-mathematical-operations/matrix-operations.py:64]:
```suggestion
# Raise the field to power 2
```

File: examples/01-mathematical-operations/matrix-operations.py [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+stress_1 = maths.pow_fc(fields_container=stress_1, factor=2.0).eval()
+
+# Add constant
+# Each component of each field is added by 2
```
Comment (PProfizi) [inline examples/01-mathematical-operations/matrix-operations.py:68]:
```suggestion
# Add a constant
# Add 2 to each value in the field
```

File: examples/01-mathematical-operations/matrix-operations.py [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+stress_2 = maths.add_constant_fc(fields_container=stress_2, ponderation=2.0).eval()
+
+# Multiply by a constant
+# Each component of each field is multiplied by 3
```
Comment (PProfizi) [inline examples/01-mathematical-operations/matrix-operations.py:72]:
```suggestion
# Multiply each value in the field by 3
```

File: examples/01-mathematical-operations/matrix-operations.py [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+stress_3 = maths.scale_fc(fields_container=stress_3, ponderation=3.0).eval()
+
+# Add fields containers
+# Each component of each field is added by the correspondent component of the others fields
```
Comment (PProfizi) [inline examples/01-mathematical-operations/matrix-operations.py:76]:
```suggestion
# Each value of each field is added by the correspondent component of the others fields
```
Here you cannot mention 'component', as it has only one. You already separated each component of the field before. Here what you are looping over is the field's entity/elementary data.

File: examples/01-mathematical-operations/matrix-operations.py [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+stress_4 = maths.add_fc(fields_container1=stress_4, fields_container2=stress_5).eval()
+stress_5 = maths.add_fc(fields_container1=stress_5, fields_container2=stress_6).eval()
+
+#Invert
```
Comment (PProfizi) [inline examples/01-mathematical-operations/matrix-operations.py:80]:
```suggestion
# Invert
```

File: examples/01-mathematical-operations/matrix-operations.py [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+# Compute the invert of each element of each field (1./X)
+stress_6 = maths.invert_fc(fields_container=stress_6).eval()
+
+################################################################################
+# Reassembling the stress tensor
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+# There are different methods to re-assemble the components
+
+# 1) With the class :class:'assemble_scalars_to_matrices_fc <ansys.dpf.core.operators.utility.assemble_scalars_to_matrices_fc.assemble_scalars_to_matrices_fc>'
```
Comment (PProfizi) [inline examples/01-mathematical-operations/matrix-operations.py:90]: Now there is only one

Comment (PProfizi) [inline examples/01-mathematical-operations/matrix-operations.py:90]:
```suggestion
# There are different methods to re-assemble the components

# 1) With the operator :class:'assemble_scalars_to_matrices_fc <ansys.dpf.core.operators.utility.assemble_scalars_to_matrices_fc.assemble_scalars_to_matrices_fc>'
```

**Review by PProfizi — COMMENTED** [review]
Very nice example, just a few things to change

File: examples/01-mathematical-operations/matrix-operations.py [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+# Printing the model displays this metadata:
+#
+# - Analysis type
+# - Available results
+# - Size of the mesh
+# - Number of results
+#
+my_model = dpf.Model(examples.find_complex_rst())
+my_mesh = my_model.metadata.meshed_region
+# print(my_model)
```
Comment (luisaFelixSalles) [inline examples/01-mathematical-operations/matrix-operations.py:35]:
```suggestion
print(my_model)
```

File: examples/01-mathematical-operations/matrix-operations.py [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+###############################################################################
+# Import the ``ansys.dpf.core`` module, included examples file, and the ``DpfPlotter``
+# module.
+from ansys.dpf import core as dpf
+from ansys.dpf.core import examples
+import ansys.dpf.core.operators.math as maths
+
+###############################################################################
+# Open an example and print the ``Model`` object. Here a result file from a crankshaft
+# under load simulation is used.
```
Comment (luisaFelixSalles) [inline examples/01-mathematical-operations/matrix-operations.py:21]:
```suggestion
```

File: examples/01-mathematical-operations/matrix-operations.py (hunk 8 above)
Comment (luisaFelixSalles) [reply examples/01-mathematical-operations/matrix-operations.py:45]:
```suggestion
print(my_avg_stress, my_avg_stress[0])
```

File: examples/01-mathematical-operations/matrix-operations.py [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+
+###############################################################################
+# Import the ``ansys.dpf.core`` module, included examples file, and the ``DpfPlotter``
+# module.
+from ansys.dpf import core as dpf
+from ansys.dpf.core import examples
+import ansys.dpf.core.operators.math as maths
+
+###############################################################################
+Open an example and print the ``Model`` object
```
Comment (luisaFelixSalles) [inline examples/01-mathematical-operations/matrix-operations.py:20]:
```suggestion
# Open an example and print the ``Model`` object
```

File: examples/01-mathematical-operations/matrix-operations.py (hunk 13 above)
Comment (luisaFelixSalles) [reply examples/01-mathematical-operations/matrix-operations.py:62]:
```suggestion
# Raise each value of the field to power 2
```

File: examples/01-mathematical-operations/matrix-operations.py [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+# Here, the stress tensor has 6 components per elementary data (symmetrical tensor XX,YY,ZZ,XY,YZ,XZ).
+
+for i in range(0, 6):  # Separating the results in different fields containers for each stress tensor component
+    globals()[f'stress_{i + 1}'] = my_avg_stress.select_component(i)
+
+################################################################################
+# Mathematical operation on each field
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+# Here we will do some basic mathematical operations on
```
Comment (luisaFelixSalles) [inline examples/01-mathematical-operations/matrix-operations.py:60]:
```suggestion
# Here we will do some basic mathematical operations on each stress field 
```

File: examples/01-mathematical-operations/matrix-operations.py [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+
+# There are different methods to re-assemble the components, here we use the
+# operator :class:'assemble_scalars_to_matrices_fc <ansys.dpf.core.operators.utility.assemble_scalars_to_matrices_fc.assemble_scalars_to_matrices_fc>'
+
+re_assemble = dpf.operators.utility.assemble_scalars_to_matrices_fc(
+    xx=stress_1, yy=stress_2, zz=stress_3, xy=stress_4, yz=stress_5, xz=stress_6, symmetrical=True
+).eval()
+
+print(re_assemble)
+print(re_assemble[0])
```
Comment (PProfizi) [inline examples/01-mathematical-operations/matrix-operations.py:124]:
```suggestion
print(re_assemble[0])

```

Comment (PProfizi) [reply examples/01-mathematical-operations/matrix-operations.py:124]: Code quality usually requests a blank line at the end of Python files. That's a convention.

**Review by PProfizi — APPROVED** [review]

**Pull request merged**